src/test/osd/test_scrubber_be.cc

```diff
@@ -0,0 +1,647 @@
+  ScrubGenerator::MockLog logger;
+  ScrubGenerator::pool_conf_t pool_conf;
+  ScrubGenerator::RealObjsConf real_objs;
+  int num_osds{0};
+
+  // ctor & initialization
+
+  TestTScrubberBe() = default;
+  ~TestTScrubberBe() = default;
+  void SetUp() override;
+  void TearDown() override;
+
+  /**
+   * Create the set of scrub-maps supposedly sent by the replica (or
+   * generated by the Primary). Then - create the snap-sets for all
+   * the objects in the set.
+   */
+  void fake_a_scrub_set(ScrubGenerator::RealObjsConfList& all_sets);
+
+  std::unique_ptr<TestScrubBackend> sbe;
+
+  spg_t spg;
+  pg_shard_t i_am;  // set to 'my osd and no shard'
+  std::set<pg_shard_t> acting_shards;
+  std::vector<int> acting_osds;
+  int acting_primary;
+
+  std::unique_ptr<TestScrubber> test_scrubber;
+
+  int64_t pool_id;
+  pg_pool_t pool_info;
+
+  OSDMapRef osdmap;
+
+  std::shared_ptr<PGPool> pool;
+  pg_info_t info;
+
+  std::unique_ptr<TestPg> test_pg;
+
+  // generated sets of "objects" for the active OSDs
+  ScrubGenerator::RealObjsConfList real_objs_list;
+
+ protected:
+  /**
+   * Create the OSDmap and populate it with one pool, based on
+   * the pool configuration.
+   * For now - only replicated pools are supported.
+   */
+  OSDMapRef setup_map(int num_osds, const ScrubGenerator::pool_conf_t& pconf);
+
+  /**
+   * Create a PG in the one pool we have. Fake the PG info.
+   * Use the primary of the PG to determine "who we are".
+   *
+   * \returns the PG info
+   */
+  pg_info_t setup_pg_in_map();
+};
+
+
+// ///////////////////////////////////////////////////////////////////////////
+// ///////////////////////////////////////////////////////////////////////////
+
+void TestTScrubberBe::SetUp()
+{
+  std::cout << "TestTScrubberBe::SetUp()\n";
+  logger.err_count = 0;
+
+  // fetch test configuration
+  auto params = inject_params();
+  pool_conf = params.pool_conf;
+  real_objs = params.objs_conf;
+  num_osds = params.num_osds;
+
+  // create the OSDMap
+
+  osdmap = setup_map(num_osds, pool_conf);
+
+  std::cout << "osdmap: " << *osdmap << std::endl;
+
+  // extract the pool from the osdmap
+
+  pool_id = osdmap->lookup_pg_pool_name(pool_conf.name);
+  const pg_pool_t* ext_pool_info = osdmap->get_pg_pool(pool_id);
+  pool =
+    std::make_shared<PGPool>(osdmap, pool_id, *ext_pool_info, pool_conf.name);
+
+  std::cout << "pool: " << pool->info << std::endl;
+
+  // a PG in that pool?
+  info = setup_pg_in_map();
+  std::cout << fmt::format("PG info: {}\n", info);
+
+  real_objs_list =
+    ScrubGenerator::make_real_objs_conf(pool_id, real_objs, acting_osds);
+
+  // now we can create the main mockers
+
+  // the "PgScrubber"
+  test_scrubber = std::make_unique<TestScrubber>(spg, osdmap, &logger);
+
+  // the "PG" (and its backend)
+  test_pg = std::make_unique<TestPg>(pool, info, i_am);
+  std::cout << fmt::format("{}: test_pg: OK\n", __func__);
+  std::cout << fmt::format("{}: acting: {}\n", __func__, acting_shards);
```

Comment:
std::endl;  std::cout doesn't necessarily flush without a std::endl, and readers of this debugging will want to be able to rely on log output ordering.

Reply:
Will change (I was opting for flushing on selected lines, just to declutter the code).